teps/0013-graceful-sidecar-termination.md

```diff
@@ -0,0 +1,226 @@
+	- Step 3: Helm Apply/Deploy
+	- Step 4: E2E
+		- Generates E2E artifacts
+		- Exposes pod and container metadata using downwards API
+
+	- Sidecar 1
+
+		The author of the task adds `Sidecar 1` that does the following:
+		- Polls the status of `Step 4` using downwards API.
+		- When the `Step 4` container is terminated, regardless of the reason of termination (failed when e2e fails and completed when e2e passes), it process all the artifacts generated by `Step 4` (Generates HTML report, does trend building etc.) and uploads the resultant artifacts to a permanent storage (S3).
+
+However, due to the current design, it terminates the sidecar immediately in case the E2E step fails. I.e., the `Step 4` container is terminated without success. 
+
+### Goals
+
+The goal of this proposal is to provide users more control over their sidecar container termination and make Tekton more flexible for complex implementation
+scenarios without introducing any breaking changes.
+
+The Nop Image termination can still be considered a good feature since in most cases, the sidecar does not need to run more than the pipelines lifetime and
+a non optimized injected sidecar may hang the pipeline indefinitely.
+
+### Non-Goals
+This proposal is not introducing any breaking change to current behavior. Sidecars still terminate using Nop Image by default when the pipeline is completed/failed.
+
+## Proposal
+
+The proposal is that we’ll make it a “per-sidecar” configuration on whether Tekton should determine and force terminate the sidecars once task run is done. I.e.; all tekton steps in the task are completed. It will default to `false` so that the new mechanism will only take effect when explicitly set by the flag in Sidecar configuration.
+
+
+### User Stories
+
+A `Pipeline` author can add a Sidecar with `waitForSidecar` enabled (true) and expect `natural death` of his/her sidecar.
+
+## Design Details
+
+A `Task` can have multiple sidecar defined each behaving according to the flag. Injected sidecars are considered as `waitForsidecar` false since that's the
+default behavior.
+
+Example:
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: Task
+spec:
+  sidecars:
+  - env:
+    - name: POD_NAME
+      valueFrom:
+        fieldRef:
+          fieldPath: metadata.name
+    - name: E2E_CONTAINER_NAME
+      value: step-e2e
+    - name: NAMESPACE
+      value: test
+    - name: BUILD_NUMBER
+      value: "25"
+    waitForSidecar: true
+    image: rifatshahnewaz/e2e-uploader:prod
+    name: publish-e2e-report
+  - env:
+    - name: POD_NAME
+      valueFrom:
+        fieldRef:
+          fieldPath: metadata.name
+    - name: LOG_STREAMER_CONTAINER_NAME
+      value: step-log-streamer
+    - name: NAMESPACE
+      value: test
+    - name: BUILD_NUMBER
+      value: "25"
+    waitForSidecar: false
+    image: rifatshahnewaz/log-streamer:prod
+    name: log-streamer
+
+    volumeMounts:
+    - mountPath: /coverage-results
+      name: e2e-report
+```
+
+Golang Example:
```

Comment:
I'm not sure this adds much that isn't already covered by the yaml example.  Is there anything specific here you're trying to demonstrate wrt the proposed behavior or syntax?

Reply:
Not really. It's just there as an example.